File: train.py

# TASK W

def devideArray(arr, size):
    result = []

    for i in range(0, len(arr), size):
        result.append(arr[i:i + size])

    return result


# Test
print(devideArray([1, 2, 3, 4, 5, 6, 7, 8, 9], 4))


# TASK V

# def countLetters(word):
#     result = {}

#     for letter in word:
#         if letter in result:
#             result[letter] += 1
#         else:
#             result[letter] = 1

#     return result


# print(countLetters("restaurantController"))


# TASK T
# def mergeSortedArrays(arr1, arr2):
#     return sorted(arr1 + arr2)


# print(mergeSortedArrays([0, 3, 4], [4, 6, 7]))


# TASK S

# def missingNumber(arr):
#     n = len(arr)
#     for i in range(n + 1):
#         if i not in arr:
#             return i


# arr1 = [3, 0, 1, 4]

# print(missingNumber(arr1))


# TASK R
# def calculate(expr):
#     a, operator, b = expr.split(" ")
#     if operator == "+":
#         return int(a) + int(b)
#     if operator == "-":
#         return int(a) - int(b)
#     if operator == "*":
#         return int(a) * int(b)
#     if operator == "/":
#         return int(a) / int(b)
#     raise ValueError("Invalid operator")


# task Q
# def hasProperty(obj, objKey):
#     for key in obj:
#         if key == objKey:
#             print(f"object has -> {objKey} <- property")
#             return True
#     print(f"object does not have -> {objKey} <- property")
#     return False


# car = {"name": "BMW", "year": 2004}

# print(hasProperty(car, "origin"))
# print(hasProperty(car, "name"))


# task P
# def objectToArray(obj):
#     result = []
#     for key in obj:
#         result.append([key, obj[key]])
#     return result


# print(objectToArray({"Name": "Alex", "age": 22, "d": 234, "group": "A2"}))


# task O
# def calculateSumOfNumbers(arr):
#     total = 0

#     for item in arr:
#         if type(item) == int or type(item) == float:
#             total += item

#     return total


# print(calculateSumOfNumbers([True, 25, 4.6, 28]))


# Task N
# def palindromCheck(text):
#     text1 = text.lower()[::-1]
#     return text.lower() == text1


# result1 = palindromCheck("pyhon")
# result2 = palindromCheck("ddddddddddddd")
# print(result2)


# TASK M
# def numberSquare(n):
#     return list(map(lambda num: {
#         "number": num,
#         "square": num ** 2
#     }, n))


# numbers = list(num for num in range(10))

# print(numberSquare(numbers))


# def reverseSentence(str):
#     str = str.lower()
#     list1 = str.split(" ")
#     print(list1)
#     reversed_words = []
#     for word in list1:
#         reversed_words.append(word[::-1])

#     return " <--> ".join(reversed_words)


# print(reverseSentence("we like coding!"))
# print(reverseSentence("Burak project"))

File: src/train.ts
/*  Project Standards:
  - Logging standards
  - Naming standards:
        function, method, variable => camelCase
        class => PascalCase
        folder => kebab-case
        css => snake_case
   - Error handling
 */

/* Cookies:
request join
self destroy */

/* Validation:
Frontend validation
Backend validation
DB validaation */

/*
  Traditional Frontend Development => SSR (Admin ) => EJS
  Modern Frontend Development => SPA (User) => REACT
*/

// TASK X

function countOccurrences(obj: any, key: string): number {
  let count = 0;

  for (const k in obj) {
    if (k === key) {
      count++;
    }

    if (typeof obj[k] === "object" && obj[k] !== null) {
      count += countOccurrences(obj[k], key);
    }
  }

  return count;
}

const data = { model: "A", s: { model: "B" }, d: { model: "F" }, };

console.log(countOccurrences(data, "model"));


// TASK W

// function divideArray(arr: number[], size: number): number[][] {
//   const result: number[][] = [];

//   for (let i = 0; i < arr.length; i += size) {
//     result.push(arr.slice(i, i + size));
//   }

//   return result;
// }
// console.log(divideArray([1, 2, 3, 4, 5, 6, 7, 8, 9], 5));

//TASK V

// function countLetters(word: string): Record<string, number> {
//   const result: Record<string, number> = {};

//   for (const letter of word) {
//     if (letter in result) {
//       result[letter]++;
//     } else {
//       result[letter] = 1;
//     }
//   }
//   console.log(result);
//   return result;
// }

// countLetters("updateChosenUser")

// // TASK T
// function mergeSortedArrays(arr1: number[], arr2: number[]): number[] {
//   const added = arr1.concat(arr2);
//   return added.sort((a, b) => a - b);
// }

// const result = mergeSortedArrays([3, 56, 7], [9, 8, 1])
// console.log(result);

// TASK S
// function missingNumber(arr: number[]): number {
//   const n = arr.length;
//   for (let i = 0; i <= n; i++) {
//     if (!arr.includes(i)) {
//       return i;
//     }
//   }
//   return -1;
// }

// const arr1 = [3, 0, 1, 4, 2, 6, 7];

// console.log(missingNumber(arr1));

// TASK R
// function calculate(expr: string): number {
//   const [a, operator, b] = expr.split(" ");
//   if (operator === "+") return Number(a) + Number(b);
//   if (operator === "-") return Number(a) - Number(b);
//   if (operator === "*") return Number(a) * Number(b);
//   if (operator === "/") return Number(a) / Number(b);
//   throw new Error("Invalid operator");
// }

// console.log(calculate("1 + 3"));
// console.log(calculate("5 * 2"));
// console.log(calculate("5 - 2"));



// TASK Q
// function hasProperty(obj: Record<string, any>, objKey: string): boolean {
//   for (const key in obj) {
//     if (key === objKey) {
//       console.log(`object has -> ${objKey} <- property`);
//       return true;
//     }
//   }
//   console.log(`object does not have -> ${objKey} <- property`);
//   return false;
// }

// const car = {
//   name: "BMW",
//   year: 2004,
// };

// console.log(hasProperty(car, "name"));
// console.log(hasProperty(car, "origin"));
// console.log(hasProperty(car, "year"));


//TASK P
// function objectToArray(obj: Record<string, any>): [string, any][] {
//   const result: [string, any][] = [];
//   for (const key in obj) {
//     result.push([key, obj[key]]);
//   }
//   return result;
// }

// console.log(objectToArray({ name: "Alex", age: 22, c: 30 }));




// task O
// function calculateSumOfNumbers(arr: any[]): number {
//   let total = 0;

//   for (const item of arr) {
//     if (typeof item === "number") {
//       total += item;
//     }
//   }

//   return total;
// }

// console.log(calculateSumOfNumbers([1, "10", { son: 10 }, true, 60, 29]));











//Task N
// function findPalindrom(str: String): boolean {
//     const str1: string = str.toLowerCase().split('').reverse().join('');
//     return str == str1;
// }
// const result1: boolean = findPalindrom("typescript")
// const result2: boolean = findPalindrom("daaad")
// console.log(result1)
// console.log(result2)




//task M
// function numberSquare(n: number[]) {
//     const result = n.map((num) => ({
//         number: num,
//         square: num ** 2,
//     }));
//     return result;
// }

// const numbers = [1, 2, 3, 4, 5, 6];

// console.log(numberSquare(numbers));